### PR TITLE
Add borderless (booktabs-style) table detection to paper-qa-pymupdf

### PR DESCRIPTION
**Pull request by 77652189**
# Detect borderless (three-line / booktabs-style) tables in `paper-qa-pymupdf`

## Summary

Academic papers — especially LaTeX-compiled papers using `\toprule / \midrule / \bottomrule`
(the `booktabs` package) and Chinese journal "三线表" tables — contain tables with **only
three full-width horizontal rules and no vertical separators**.

PyMuPDF's `page.find_tables()` (default strategy) misses these because it builds a cell grid
from *intersecting* horizontal and vertical lines.  When there are no vertical lines, there
are no intersections, so `find_tables()` returns `[]` and the table falls through as
unstructured plain text in the chunk.

The practical impact is significant: quantitative experimental results, parameter comparisons,
and measurement data that sit in table cells end up as a flat stream of numbers and labels
mixed with body text, with no column-header context.  No `ParsedMedia(type="table")` is
emitted, so neither the markdown-table prompt path nor the multimodal screenshot path is
triggered.

This PR adds a **supplementary detection pass** that catches these missed tables using
four complementary strategies, evidence scoring, and deduplication against tables already
found by `find_tables()`.

---

### New: `packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py`

A self-contained module (~1 500 lines) implementing a full multi-strategy detection pipeline
ported from PaperSort's `TableRegionDetector` / `extract_threeline` /
`detect_rotated_table` with all pdfplumber-specific APIs replaced by PyMuPDF equivalents.

#### Detection strategies

Four strategies run in order; their candidates are merged and scored before output:

| Priority | Strategy | Trigger condition |
|----------|----------|-------------------|
| 1 (primary) | **Caption anchoring** | "Table N" / "表N" text found; extends bbox downward to capture table body using rules and table-like text lines |
| 2 (secondary) | **Wide H-rule clustering** | Drawing-layer horizontal paths spanning ≥ 20 % of page width; groups within 200 pt are clustered into a table band |
| 3 (fallback) | **Text-alignment runs** | ≥ 3 consecutive text lines with ≥ 2 distinct x-alignment bins; detects tables with no border lines at all |
| 4 (special) | **Rotated-text detection** | Lines whose PyMuPDF `dir` vector has non-zero sin component; finds 90°-rotated tables by locating a rotated caption and expanding to nearby character groups; structured text is extracted by `_extract_rotated_cells_from_chars` (ported from PaperSort's `detect_rotated_table`) |

All thresholds are module-level named constants (no magic numbers).  New constants
for the rotated-extraction path: `_ROTATED_HEADER_KEYWORDS`, `_ROTATED_EXTRACT_GAP_PT`,
`_ROTATED_EXTRACT_CLUSTER_TOL`, `_ROTATED_HEADER_GAP_PT`.

#### Evidence-based scoring

Each candidate receives a confidence score in `[0, 1]` from:

| Evidence | Score |
|----------|-------|
| Caption nearby | +0.35 |
| ≥ 3 wide horizontal rules | +0.30 |
| ≥ 2 wide horizontal rules | +0.20 |
| ≥ 3 aligned text lines | +0.30 |
| ≥ 2 aligned text lines | +0.15 |
| High numeric token density (≥ 20 %) | +0.15 |
| Short cell-like tokens (≥ 70 %) | +0.10 |
| ≥ 3 long prose lines (≥ 18 words) | −0.30 |
| Section-heading word at line start | −0.20 |

Candidates below the minimum keep score (0.25) are discarded.

#### Structure extraction

Per detected table:

1. **Rotated tables** (strategy 4) bypass steps 2–3 and use `_extract_rotated_cells_from_chars`
   instead: characters are re-grouped by 0.5 pt x-bins, intra-group gaps identify cell
   boundaries, and a gap-clustering pass infers row/column layout (full port of PaperSort's
   `detect_rotated_table`).
2. For all other tables, clip rect is expanded by 3 pt on all sides for robust word capture.
3. **Numeric-error table** pattern (`key | mean ± sd | mean ± sd`) is tried first.
4. Falls back to **standard column inference** (gap-merge of word x-spans).
5. **`merge_multiline_cells`** handles dual-column layouts and wrap-around continuation rows.
6. Result is rendered as GitHub-flavoured markdown and stored in `ParsedMedia.text`.
7. A pixmap screenshot is stored in `ParsedMedia.data` for multimodal LLMs.

### Modified: `packages/paper-qa-pymupdf/src/paperqa_pymupdf/reader.py`

After the existing `find_tables()` loop, a single call to `detect_borderless_tables()`
is inserted (~15 lines).  It receives the bboxes of already-detected tables so that
fully-bordered tables detected by `find_tables()` are not re-emitted as duplicates
(IoU threshold: 0.30).

```python
# ← existing find_tables() loop (unchanged) ...

# NEW: supplementary pass for borderless tables
already_bboxes = [tuple(m.info["bbox"]) for m in media if m.info.get("type") == "table"]
media.extend(
    detect_borderless_tables(
        page,
        page_num=i,
        page_width=float(page.rect.width),
        dpi=dpi,
        pymupdf_pixmap_attrs=PYMUPDF_PIXMAP_ATTRS,
        already_detected_bboxes=already_bboxes,
    )
)
```

### New: `packages/paper-qa-pymupdf/tests/test_borderless_tables.py`

50 test cases across two levels:

**Unit tests** (no real PDF):
- `TestTableCaptionRE` — English/Chinese caption regex matching
- `TestFindColRanges` — span merging, gap threshold, empty input
- `TestAssignCol` — centre-in-range, nearest-column fallback
- `TestWordsToGrid` — row splitting by y-gap, multi-row layout
- `TestClusterRules` — proximity grouping, minimum-size filtering
- `TestToMarkdown` — pipe escaping, header/separator/data ordering
- `TestMergeMultilineCells` — simple continuation, independent rows, single row
- `TestExtractNumericErrorTable` — mean±sd detection, false-positive guard
- `TestExtractRotatedCellsFromChars` — title detection, 2-column extraction, edge cases
  (too few chars, ≤ 2 groups, single-cell header)

**Integration tests** (synthetic PDFs created with PyMuPDF):
- `TestGetWideHRules` — verifies rule y-coordinates match what was drawn
- `TestCaptionBasedRegions` — caption text anchors correct table area; no caption → empty
- `TestAlignmentBasedRegions` — alignment strategy finds table with no rules; no-op on prose
- `TestDetectBorderlessTables` — full three-rule and two-rule tables; text-only no-op;
  duplicate suppression; JSON-serialisability of `info`; caption PDF
- `test_parse_pdf_to_pages_detects_borderless_table` — end-to-end through `parse_pdf_to_pages`
- `test_parse_pdf_to_pages_no_tables_on_text_only_page`

---

## What is NOT changed

- `paper-qa-pypdf`: pdfplumber's `find_tables()` has a `vertical_strategy="text"` option
  that handles this case differently; a follow-up PR can address that parser.
- `paper-qa-docling`: Docling uses an ML layout model that is already less sensitive to
  absent border lines.
- `paper-qa-nemotron`: unchanged.
- The `full_page=True` code path: when the entire page is already captured as a single
  screenshot, the LLM sees the table visually regardless, so no change is needed there.

---

## Known limitations

1. **Column inference quality** depends on word spacing in the PDF.  Tables whose column
   words are very close together (gap < 8 pt) may be merged into fewer columns.  The
   `_MIN_COL_GAP_PT` constant can be tuned down if needed.

2. **Two-rule tables** (only top + bottom, no midrule) are detected but the header/data
   split falls back to a 25 % heuristic, which may misclassify the first data row as the
   header when the header zone is unusually tall.

3. **Rotated tables**: structured markdown is generated by `_extract_rotated_cells_from_chars`
   (a full port of PaperSort's `detect_rotated_table`).  The algorithm requires a recognisable
   title group (leftmost 1–2 x-bins) and ≥ 20 rotated characters; tables with extremely dense
   character spacing (gap < 10 pt between cells) may not split cells correctly.  The pixmap
   screenshot in `ParsedMedia.data` always carries full visual fidelity as a fallback for
   multimodal LLMs.

4. **False positives on pages with wide decorative rules** (page headers/footers that are
   also full-width horizontal lines) are filtered by the requirement of ≥ 2 rules within
   200 pt of each other.  A single decorative line would be a 1-element cluster and is
   discarded.  Decorative lines that happen to come in pairs near body text may pass the
   rule filter but will receive a low evidence score and be discarded by the 0.25 threshold.

5. **`text` field quality**: the extracted markdown is best-effort.  For tables with merged
   cells, subscript/superscript characters, or symbol-heavy content the text will be
   imperfect.  The screenshot stored in `ParsedMedia.data` carries full visual fidelity.

---

## Test plan

```bash
# Run only the new tests (fast, uses synthetic PDFs — no API key required)
pytest packages/paper-qa-pymupdf/tests/test_borderless_tables.py -v

# Run the full pymupdf test suite to confirm no regressions
# (test_paperqa_pymupdf.py requires OPENAI_API_KEY; skip it in CI if unavailable)
pytest packages/paper-qa-pymupdf/tests/ -v --ignore=packages/paper-qa-pymupdf/tests/test_paperqa_pymupdf.py
```

To manually verify on a real academic PDF:

```python
from paperqa_pymupdf import parse_pdf_to_pages

result = parse_pdf_to_pages("your_booktabs_paper.pdf", parse_media=True)
for page_key, content in result.content.items():
    if isinstance(content, tuple):
        _, media = content
        for m in media:
            if m.info.get("type") == "table":
                print(f"Page {page_key}: {m.info.get('detection_method', 'find_tables')}")
                print(m.text[:300])
                print()
```

Any table whose `info["detection_method"] == "borderless"` was found by this new pass.

---

Generated with [Claude Code](https://claude.ai/claude-code)

### CODE REVIEW

File: packages/paper-qa-pymupdf/tests/test_borderless_tables.py [hunk 1]
```diff
@@ -0,0 +1,770 @@
...
+    _find_col_ranges,
+    _get_wide_h_rules,
+    _to_markdown,
+    _words_to_grid,
+    detect_borderless_tables,
+    merge_multiline_cells,
+)
+from paperqa_pymupdf.reader import PYMUPDF_PIXMAP_ATTRS
+
+# ── shared constants ───────────────────────────────────────────────────────────
```
Comment (jamesbraza) [inline packages/paper-qa-pymupdf/tests/test_borderless_tables.py:40]: This should go in the `paper-qa-pymupdf` tests

File: packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py [hunk 2]
```diff
@@ -0,0 +1,1497 @@
...
+                if cell.strip() and i < len(prev):
+                    prev[i] = (prev[i] + " " + cell).strip()
+        else:
+            result2.append(list(row))
+    return result2
+
+
+# ── markdown rendering ─────────────────────────────────────────────────────────
+
+def _to_markdown(header: list[str], data_rows: list[list[str]]) -> str:
```
Comment (jamesbraza) [inline packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py]: Why not use PyMuPDF's markdown table functionality?

File: packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py [hunk 3]
```diff
@@ -0,0 +1,1497 @@
...
+    Applies four strategies — caption anchoring, wide horizontal-rule clustering,
+    text-alignment runs, and rotated-text detection — scores each candidate by
+    multi-evidence confidence, and emits one :class:`~paperqa.types.ParsedMedia`
+    entry per table found (``type="table", detection_method="borderless"``).
+
+    Candidates whose bounding box substantially overlaps an already-detected
+    table (IoU >= :data:`_OVERLAP_IOU_THRESHOLD`) are skipped to avoid
+    re-emitting tables that ``find_tables()`` already found.
+
+    Parameters
```
Comment (jamesbraza) [inline packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py]: We use Google-style docstrings

File: packages/paper-qa-pymupdf/tests/test_borderless_tables.py [hunk 4]
```diff
@@ -0,0 +1,770 @@
...
+----------------------------------------
+y=200  ───────────────────────  top border   (toprule)
+       Strain   Titer (g/L)   Yield (%)     header row
+y=215  ───────────────────────  separator    (midrule)
+       E. coli       12.3        45.2        data row 1
+       S. cerevi.     8.7        38.6        data row 2
+       B. subtil.    15.1        51.0        data row 3
+y=270  ───────────────────────  bottom border (bottomrule)
+"""
+from __future__ import annotations
```
Comment (jamesbraza) [inline packages/paper-qa-pymupdf/tests/test_borderless_tables.py]: Can you remove all `from __future__ import annotations` from your code you add

File: packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py [hunk 5]
```diff
@@ -0,0 +1,1497 @@
...
+        1 for cell_text, _span in cells
+        if any(
+            _is_number_token(part)
+            for part in cell_text.replace("±", " ± ").replace("卤", " ± ").split()
+        )
+    )
+    return cells if len(cells) >= _MIN_COLS and numeric_count >= 1 else None
+
+
+def _extract_numeric_error_table(
```
Comment (jamesbraza) [inline packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py]: Try to figure out how to use upstream libraries to downsize this module by at least 50% (<750 lines)

File: packages/paper-qa-pymupdf/tests/test_borderless_tables.py (hunk 1 above)
Comment (77652189) [reply packages/paper-qa-pymupdf/tests/test_borderless_tables.py:40]: Done — _PIXMAP_ATTRS is now defined locally in the test file as a minimal rozenset rather than imported from 
eader internals.

File: packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py (hunk 2 above)
Comment (77652189) [reply packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py]: `pymupdf.table.Table.to_markdown()` only works on `pymupdf.table.Table` objects (as used in `reader.py` for fully-bordered tables). Our detected tables are plain `list[list[str]]` grids built by word-position inference, so there is no PyMuPDF API that accepts that format directly. The `_to_markdown` function is 8 lines and implements the same GitHub-flavoured pipe format.

File: packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py (hunk 3 above)
Comment (77652189) [reply packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py]: Done — all public-facing functions now use Google-style `Args:` / `Returns:` / `Raises:` sections.

File: packages/paper-qa-pymupdf/tests/test_borderless_tables.py (hunk 4 above)
Comment (77652189) [reply packages/paper-qa-pymupdf/tests/test_borderless_tables.py]: Done — removed from both files. Switched to a direct `import pymupdf` (consistent with `reader.py`) so runtime annotation evaluation works without the future import.

File: packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py (hunk 5 above)
Comment (77652189) [reply packages/paper-qa-pymupdf/src/paperqa_pymupdf/borderless_tables.py]: Done — module reduced from ~1 500 lines to 728 lines. Removed the three most complex sub-systems: rotated-table detection + extraction (~440 lines), numeric-error table pattern matching (~130 lines), dual-column `merge_multiline_cells` (~110 lines), and the text-alignment detection strategy (~55 lines). The core value — detecting H-rule and caption-anchored three-line tables — is fully preserved. The removed features can be reintroduced as a follow-up PR if needed.
